File: web/client/src/protocol.ts
export type RoomPhase = "waiting" | "playing" | "finished";
export type GameKind = "on_mars";

export type LobbyRoomSummary = {
  id: string;
  name: string;
  host_nickname: string;
  player_count: number;
  max_players: number;
  game_kind: GameKind;
};

export type PlayerPayload = {
  id: string;
  nickname: string;
  is_host: boolean;
  color: string | null;
};

export type RoomStatePayload = {
  id: string;
  name: string;
  phase: RoomPhase;
  host_id: string;
  players: PlayerPayload[];
  max_players: number;
  game_kind: GameKind;
};

export type OnMarsStubPlayer = {
  id: string;
  nickname: string;
  color: string;
};

/** Stub WS jusqu’au branchement de la session de test. */
export type OnMarsGameState = {
  kind: "on_mars";
  state: {
    players: OnMarsStubPlayer[];
    stub: boolean;
  };
};

export type GameStatePayload = OnMarsGameState;

export type ClientMessage =
  | { type: "set_nickname"; nickname: string }
  | { type: "create_room"; name: string; game_kind: GameKind; max_players: number }
  | { type: "join_room"; room_id: string }
  | { type: "leave_room" }
  | { type: "start_game" }
  | { type: "rematch" }
  | { type: "back_to_lobby" }
  | { type: "reconnect"; player_id: string }
  | { type: "rtc_signal"; target: string; payload: unknown };

export type ServerMessage =
  | { type: "welcome"; player_id: string }
  | { type: "lobby_list"; rooms: LobbyRoomSummary[] }
  | { type: "room_state"; room: RoomStatePayload }
  | { type: "game_state"; game: GameStatePayload }
  | { type: "error"; code: string; message: string }
  | { type: "rtc_signal"; from: string; payload: unknown };

File: web/server/src/protocol.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::game::{GameKind, OnMarsStub};
use crate::state::RoomPhase;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    SetNickname { nickname: String },
    CreateRoom {
        name: String,
        #[serde(default = "default_on_mars")]
        game_kind: GameKind,
        #[serde(default = "default_max")]
        max_players: usize,
    },
    JoinRoom { room_id: Uuid },
    LeaveRoom,
    StartGame,
    Rematch,
    BackToLobby,
    Reconnect { player_id: Uuid },
    RtcSignal { target: Uuid, payload: serde_json::Value },
}

fn default_on_mars() -> GameKind {
    GameKind::OnMars
}
fn default_max() -> usize {
    2
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    Welcome {
        player_id: Uuid,
    },
    LobbyList {
        rooms: Vec<LobbyRoomSummary>,
    },
    RoomState {
        room: RoomStatePayload,
    },
    GameState {
        game: GameStatePayload,
    },
    Error {
        code: String,
        message: String,
    },
    RtcSignal {
        from: Uuid,
        payload: serde_json::Value,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LobbyRoomSummary {
    pub id: Uuid,
    pub name: String,
    pub host_nickname: String,
    pub player_count: usize,
    pub max_players: usize,
    pub game_kind: GameKind,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerPayload {
    pub id: Uuid,
    pub nickname: String,
    pub is_host: bool,
    pub color: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomStatePayload {
    pub id: Uuid,
    pub name: String,
    pub phase: RoomPhase,
    pub host_id: Uuid,
    pub players: Vec<PlayerPayload>,
    pub max_players: usize,
    pub game_kind: GameKind,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum GameStatePayload {
    OnMars {
        state: OnMarsStub,
    },
}
